### ISSUE TITLE
New buzzwords How Deep Learning Supercharges Natural Language Processing

### ISSUE THREAD

**Issue by sudocurse** [sudocurse/reading-list#8591]
How Deep Learning Supercharges Natural Language Processing<br>
http://ift.tt/2G8esD3<br>
<img width="640" height="426" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/03/43f7607b-block-chain-3145392_1280-1024x682.jpg" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2018/03/43f7607b-block-chain-3145392_1280.jpg"> <a href="https://insightengines.com/" class="clearfix infoBlock" target="_blank"></a>
<div class="infoBlockImageBlock"><img class="infoBlockImage" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/03/533b2196-jacobperkinsheadshot.jpg" style="border-radius: 50%;"></div>
<div class="infoBlockTextBlock">
<div class="infoBlockTitle" style="display: none;"></div>
<div class="infoBlockText">Jacob Perkins is the co-founder and CTO of InsightEngines.com, creating technology to ask simple natural language questions over complex data. Currently focused on Cyber Security customers, Insight Engines helps democratize access to machine data, making it possible for anyone to perform useful queries, while helping experts save time and resources. Jacob is a contributor to the NLTK project, an open source natural language toolkit for Python, as well as the creator of nltk-trainer, which makes it easy to train your own natural language machine learning models using NLTK. He is the author of <em>Python Text Processing with NLTK Cookbook</em>, a contributor to OReilly's <em>Bad Data Handbook</em>, and sometimes writes about natural language technology at streamhacker.com.</div>
</div>
<p>Voice search, intelligent assistants, and chatbots are becoming common features of modern technology. Users and customers are demanding a better, more human experience when interacting with computers. According to <a href="https://www.tableau.com/reports/business-intelligence-trends#nlp" class="ext-link" rel="external nofollow">Tableau&rsquo;s business trends report</a>, IDC predicts that by 2019, intelligent assistants will become commonly accessible to enterprise workers, while Gartner predicts that by 2020, 50 percent of analytics queries will involve some form of natural language processing. Chatbots, intelligent assistants, natural language queries, and voice-enabled applications all involve various forms of natural language processing. To fully realize these new user experiences, we will need to build upon the latest methods, some of which I will cover here.</p>
<p>Let&rsquo;s start with the basics: what is natural language processing? Natural language processing (NLP), is a collection of techniques for helping machines understand human language. For example, one of the essential techniques is tokenization: breaking up text into &ldquo;tokens,&rdquo; such as words. Given individual words in sequence, you can start to apply reason to them, and do things like sentiment analysis to determine if a piece of text is positive or negative. But even a task as simple as word identification can be quite tricky. Is the word what&rsquo;s really one word or two (what + is, or what + was)? What about languages that use characters to represent multi-word concepts, like <a href="https://en.wikipedia.org/wiki/Kanji" class="ext-link" rel="external nofollow">Kanjii</a>?</p>
<p>Deep learning is an advanced type of machine learning using neural networks. It became popular due to the success of the techniques at solving problems such as image classification (labeling an image based on visual content) and speech recognition (converting sounds into text). Many people thought that deep learning techniques, when applied to natural language, would quickly achieve similar levels of performance. But because of all the idiosyncrasies of natural language, the field has not seen the same kind of breakthrough success with deep learning as other fields, like image processing. However, that appears to be changing. In the past few years, researchers have been applying newer deep learning methods to natural language processing, and I will share some of these recent successes.</p>
<p>Deep learning&nbsp;&mdash; through recent improvements to word embeddings, a focus on attention, mobile enablement, and its appearance in the home&nbsp;&mdash; is starting to capture natural language processing like it previously captured image processing. In this article, I will cover some recent deep learning-based NLP research successes that have made an impact on the field. Because of these improvements, we will see simpler and more natural user experiences, better software performance, and more powerful home and mobile applications.</p>
<h2>Word Embeddings</h2>
<p>Words are essential to every natural language processing system. Traditional NLP looks at words as strings, but deep learning techniques can only process numeric vectors. <a href="http://ruder.io/word-embeddings-2017/" class="ext-link" rel="external nofollow">Word embeddings</a> were invented as a way to <a href="https://monkeylearn.com/blog/word-embeddings-transform-text-numbers/" class="ext-link" rel="external nofollow">transform words into vectors</a>, enabling new kinds of mathematical feature analysis. But the vector representation of words is only as good as the text it was trained on.</p>
<p>The more common word embeddings are trained on Wikipedia, but Wikipedia text may not be representative of whatever text you&rsquo;re processing. It&rsquo;s generally written as well structured factual statements, which is nothing like text found on twitter, and both of these are different than restaurant reviews. So vectors trained on Wikipedia might be mathematically misleading if you use those vectors to analyze a different style of text. Text from the <a href="https://aws.amazon.com/public-datasets/common-crawl/" class="ext-link" rel="external nofollow">Common Crawl</a> provides a more diverse set of text for training a word embedding model. The <a href="https://fasttext.cc/" class="ext-link" rel="external nofollow">FastText</a> library provides some great pre-trained English word vectors, along with tools for training your own. Training your own vectors is essential if you&rsquo;re processing any language other than English.</p>
<p>Character level embeddings have also shown surprising results. This technique tries to learn vectors for individual characters, where words would be represented as a composition of the individual character vectors. In an effort to learn how to predict the next character in reviews, researchers discovered a <a href="https://blog.openai.com/unsupervised-sentiment-neuron/" class="ext-link" rel="external nofollow">sentiment neuron</a>, which they could control to produce positive or negative review output. Using the sentiment neuron, they were able to beat the previous top accuracy score on the sentiment treebank. This is quite an impressive result for something discovered as a side effect of other research.</p>
<h2>CNNs, RNNs, and Attention</h2>
<p>Moving beyond vectors, deep learning requires training <a href="https://en.wikipedia.org/wiki/Artificial_neural_network" class="ext-link" rel="external nofollow">neural networks</a> for various tasks. Vectors are the input and output, in between are layers of nodes connected together in a network. The nodes represent functions on the input data, with each function taking the input from the previous layer and producing output for the next layer. The structure of the network and how the nodes are connected very much determines the learning capabilities and performance.</p>
<p>In general, the deeper and more complicated a network, the longer it takes to train. When using large datasets, many networks can only be effectively trained using clusters of graphics processors (GPUs), because GPUs are optimized for the necessary floating point math. This puts some types of deep learning outside the reach of anyone not at large companies or institutions that can afford the expensive GPU clusters necessary for deep learning on big data.</p>
<p>Standard neural networks are feedforward networks, where each node in a layer is forward connected to every node in the next layer. A Recurrent Neural Network (RNN) is a network where the nodes in each layer also connect back to the previous layer. This creates a kind of memory that can be great for learning from sequences, such as words in a sentence.</p>
<p>A Convolutional Neural Networks (CNN) is a type feedforward network, but with more layers, and where the forward connections have been manipulated, or convoluted, to achieve certain properties. CNNs tend to be good at extracting position invariant features, meaning they do not care so much about sequence ordering. Because of this, CNNs can be trained in a more parallel manner, leading to faster training and optimization compared to RNNs.</p>
<p>While CNNs may win in raw speed, both types of neural networks tend to have comparable performance characteristics. &nbsp;In fact, RNNs have a slight edge when it comes to sequence oriented tasks like Part-of-Speech tagging, where you are trying to identify the part of speech (such as &ldquo;noun&rdquo; or &ldquo;verb&rdquo;) for each word in a sentence. For a detailed performance comparison of CNNs and RNNs applied to NLP see: <a href="https://arxiv.org/abs/1702.01923" class="ext-link" rel="external nofollow">Comparative Study of CNN and RNN for Natural Language Processing.</a></p>
<p>The most successful RNN models are the LSTM (Long short-term memory) and GRU (gated recurrent unit). These use attention gates, which act as a kind of short-term memory for the network. However, a newer research paper implies that <a href="https://arxiv.org/abs/1706.03762" class="ext-link" rel="external nofollow">attention may be all you need</a>. By doing away with recurrence networks and convolution, and keeping only attention mechanisms, these models can be trained in parallel like a CNN, but even faster, and have comparable better performance than RNNs on some sequence learning tasks, such machine translation.</p>
<p>Reducing the training cost while maintaining comparable performance means that smaller companies and individuals can throw more data at their deep learning models, and potentially compete more effectively with larger companies and institutions.</p>
<h2>Software 2.0</h2>
<p>One of the nice properties of neural network models is that the core algorithms and math are mostly the same. Once you have the infrastructure, model definition, and training algorithms all setup, these models are very reusable. &ldquo;<a href="https://medium.com/@karpathy/software-2-0-a64152b37c35" class="ext-link" rel="external nofollow">Software 2.0</a>&rdquo; is the idea that significant components of an application or system can be replaced by neural network models. Instead of writing code, developers:</p>
<ol>
<li>Collect training data</li>
<li>Clean and label the data</li>
<li>Train a model</li>
<li>Integrate the model</li>
</ol>
<p>While the most interesting parts are often steps three and four, most of the work happens in the data preparation steps one and two. Collecting and curating good, useful, clean data can be a significant amount of work, which is why methods like <a href="https://www.slideshare.net/japerk/corpus-bootstrapping-with-nltk" class="ext-link" rel="external nofollow">corpus bootstrapping</a> are important for getting to good data faster. In the long run, it is often easier to make better data than it is to design better algorithms.</p>
<p>The past few years have demonstrated that neural networks can achieve much better performance than many alternatives, sometimes even in areas not traditionally touched by machine learning. One of the most recent and interesting advances is in <a href="https://arxiv.org/abs/1712.01208" class="ext-link" rel="external nofollow">learning data indexing structures</a>. B-tree indexes are a commonly used data structure that provides an efficient way of finding data, assuming the tree is structured well. However, these newly learned indexes significantly outperformed the traditional B-tree indexes in both speed and memory usage. Such low-level data structure performance improvements could have far-reaching impacts if it can be integrated into standard development practices.</p>
<p>As research progresses, and the necessary infrastructure becomes cheaper and more available, deep learning models are likely to be used in more and more parts of the software stack, including mobile applications.</p>
<h2>Mobile Machine Learning</h2>
<p>Most deep learning requires clusters of expensive GPUs and lots of RAM. This level of compute power is only accessible to those who can afford it, usually in the cloud. But consumers are increasingly using mobile devices, and much of the world does not have reliable and affordable full-time wireless connectivity. Getting machine learning into mobile devices will enable more developers to create all sorts of new applications.</p>
<ul>
<li>Apple&rsquo;s <a href="https://developer.apple.com/machine-learning/" class="ext-link" rel="external nofollow">CoreML</a> framework enables a number of NLP capabilities on iOS devices, such as language identification and named entity recognition.</li>
<li>Baidu developed a CNN library for <a href="https://github.com/baidu/mobile-deep-learning" class="ext-link" rel="external nofollow">mobile deep learning</a> that works on both iOS and Android.</li>
<li>Qualcomm created a <a href="https://developer.qualcomm.com/software/snapdragon-neural-processing-engine" class="ext-link" rel="external nofollow">Neural Processing Engine</a> for its mobile processors, enabling popular deep learning frameworks to operate on mobile devices.</li>
</ul>
<p>Expect a lot more of this in the near future, as mobile devices continue to become more powerful and ubiquitous. Marc Andreessen famously said that &ldquo;<a href="https://a16z.com/2016/08/20/why-software-is-eating-the-world/" class="ext-link" rel="external nofollow">software is eating the world</a>,&rdquo; and now machine learning appears to be eating software. Not only is it in our pocket, it is also in our homes.</p>
<h2>Deep Learning in the Home</h2>
<p><a href="https://en.wikipedia.org/wiki/Amazon_Alexa" class="ext-link" rel="external nofollow">Alexa</a> and other voice assistants became mainstream in 2017, bringing NLP into millions of homes. Mobile users are already familiar with <a href="https://en.wikipedia.org/wiki/Siri" class="ext-link" rel="external nofollow">Siri</a> and <a href="https://en.wikipedia.org/wiki/Google_Assistant" class="ext-link" rel="external nofollow">Google Assistant</a>, but the popularity of Alexa and <a href="https://en.wikipedia.org/wiki/Google_Home" class="ext-link" rel="external nofollow">Google Home</a> shows how many people have become comfortable having conversations with voice-activated dialogue systems. How much these systems rely on deep learning is somewhat unknown, but it is fairly certain that significant parts of their <a href="https://arxiv.org/abs/1711.01731" class="ext-link" rel="external nofollow">dialogue systems use deep learning models</a> for core functions such as speech to text, part of speech tagging, natural language generation, and text to speech.</p>
<p>As research advances and these companies collect increasing amounts of data from their users, deep learning capabilities will improve as well, and implementations of &ldquo;software 2.0&rdquo; will become pervasive. While a few large companies are creating powerful data moats, there is always room on the edges for highly specialized, domain-specific applications of natural languages, such as cybersecurity, IT operations, and data analytics.</p>
<p>Deep learning has become a core component of modern natural language processing systems.</p>
<p>However, many traditional natural language processing techniques are still quite effective and useful, especially in areas that lack the huge amounts of training data necessary for deep learning. I will cover these traditional statistical techniques in an upcoming article.</p>
<p class="attribution">Feature image <a href="https://pixabay.com/en/block-chain-data-records-concept-3145392/" class="ext-link" rel="external nofollow">via</a> Pixabay.</p>
<p>The post <a rel="nofollow" href="https://thenewstack.io/recent-advances-deep-learning-natural-language-processing/">How Deep Learning Supercharges Natural Language Processing</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br>
<br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
March 20, 2018 at 06:00AM
